source/Calamari.Common/Features/StructuredVariables/XmlFormatVariableReplacer.cs

```diff
@@ -32,7 +33,13 @@ public bool IsBestReplacerForFileName(string fileName)
 
         public void ModifyFile(string filePath, IVariables variables)
         {
-            var fileText = fileSystem.ReadFile(filePath, out var encoding);
+            var encodingPrecedence = new List<Encoding>(CalamariPhysicalFileSystem.DefaultInputEncodingPrecedence);
+
+            var fileBytes = fileSystem.ReadAllBytes(filePath);
+            if (TryGetDeclaredEncoding(fileBytes) is {} declaredEncoding)
+                encodingPrecedence.Insert(0, declaredEncoding);
```

Comment:
This is nice. Now that we've got all the infrastructure in place to handle falling back between encoding, I love how simple it is to handle XML declarations. This could have been so much more painful!